// @angular/common/src/directives/ng_for.ngfactory.js
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register(['@angular/common/src/directives/ng_for', '@angular/core/src/change_detection/change_detection', '@angular/core/src/linker/view_utils'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var import0, import1, import2;
    var Wrapper_NgFor;
    return {
        setters:[
            function (import0_1) {
                import0 = import0_1;
            },
            function (import1_1) {
                import1 = import1_1;
            },
            function (import2_1) {
                import2 = import2_1;
            }],
        execute: function() {
            Wrapper_NgFor = (function () {
                function Wrapper_NgFor(p0, p1, p2, p3) {
                    this.changed = false;
                    this.changes = {};
                    this.context = new import0.NgFor(p0, p1, p2, p3);
                    this._expr_0 = import1.UNINITIALIZED;
                    this._expr_1 = import1.UNINITIALIZED;
                    this._expr_2 = import1.UNINITIALIZED;
                }
                Wrapper_NgFor.prototype.check_ngForOf = function (currValue, throwOnChange, forceUpdate) {
                    if ((forceUpdate || import2.checkBinding(throwOnChange, this._expr_0, currValue))) {
                        this.changed = true;
                        this.context.ngForOf = currValue;
                        this.changes['ngForOf'] = new import1.SimpleChange(this._expr_0, currValue);
                        this._expr_0 = currValue;
                    }
                };
                Wrapper_NgFor.prototype.check_ngForTrackBy = function (currValue, throwOnChange, forceUpdate) {
                    if ((forceUpdate || import2.checkBinding(throwOnChange, this._expr_1, currValue))) {
                        this.changed = true;
                        this.context.ngForTrackBy = currValue;
                        this.changes['ngForTrackBy'] = new import1.SimpleChange(this._expr_1, currValue);
                        this._expr_1 = currValue;
                    }
                };
                Wrapper_NgFor.prototype.check_ngForTemplate = function (currValue, throwOnChange, forceUpdate) {
                    if ((forceUpdate || import2.checkBinding(throwOnChange, this._expr_2, currValue))) {
                        this.changed = true;
                        this.context.ngForTemplate = currValue;
                        this.changes['ngForTemplate'] = new import1.SimpleChange(this._expr_2, currValue);
                        this._expr_2 = currValue;
                    }
                };
                Wrapper_NgFor.prototype.detectChangesInInputProps = function (view, el, throwOnChange) {
                    var changed = this.changed;
                    this.changed = false;
                    if (!throwOnChange) {
                        if (changed) {
                            this.context.ngOnChanges(this.changes);
                            this.changes = {};
                        }
                        this.context.ngDoCheck();
                    }
                    return changed;
                };
                Wrapper_NgFor.prototype.detectChangesInHostProps = function (view, el, throwOnChange) {
                };
                return Wrapper_NgFor;
            }());
            exports_1("Wrapper_NgFor", Wrapper_NgFor);
        }
    }
});
